### ISSUE TITLE
Ways to Compute the Test Perplexity

### ISSUE THREAD

**Issue by zhongpeixiang** [bentrevett/pytorch-seq2seq#146]
Thank you for sharing your great tutorial about seq2seq models.

I have one confusion regarding how the test perplexity is computed.

From my understanding, the test perplexity is `exp(loss)` where `loss` is the averaged negative log-likelihood of the groundtruth tokens. In other words, `loss = -1/N * (log(p(w_1|<sos>)) + log(p(w_2|<sos>, w_1)) + ... + log(p(w_N|<sos>, w_1, ..., w_(N-1))))`, where `w_1, ..., w_N` denote `N` groundtruth tokens. 

However, in your tutorial, the loss is calculated without teacher forcing, i.e., the current predicted token is fed to the model for the next token prediction. In this way, the loss no longer measures the probability of the groudtruth sentence. 

From my understanding, the perplexity should be independent of the decoding algorithm (greedy decoding, beam search, etc.) and calculated with teacher forcing (see [this implementation](https://github.com/OpenNMT/OpenNMT-py/blob/7cbfbb38190baa0a7b98260435c47fa6beb13461/onmt/translate/translator.py#L874)).

Now I'm a bit confused.

**Comment by bentrevett** [comment]
I only added the perplexity metric because it's a lot easier to compare perplexities than to compare loss values - going from 35 to 25 test perplexity also sounds better than going from a test loss of 3.5 to 3.2. 

Translation should definitely be evaluated using a teacher forcing ratio of zero and I don't see perplexity used for translation so I'd take the perplexity metric with a grain of salt and assume that I am miscalculating it. 

What I should be doing is calculating BLEU every epoch but I didn't want to introduce inference too early, plus BLEU takes a non-trivial amount of time.

I might completely remove the perplexity metric if you think it adds confusion.

**Comment by zhongpeixiang** [comment]
Thank you for the detailed reply.

Do you know how the test perplexity is calculated in language modeling? with or without teacher forcing?

**Comment by bentrevett** [comment]
I believe it is calculated with teacher forcing, see: https://github.com/salesforce/awd-lstm-lm/blob/master/main.py#L166

I also found [this](https://twitter.com/joeddav/status/1283079160630439936) twitter thread had some useful information for calculating perplexity in language models.

**Comment by zhongpeixiang** [comment]
Thank you for the very helpful resources!
